library/ziti.c

```diff
@@ -160,6 +184,222 @@ int load_tls(nf_config *cfg, tls_context **ctx) {
+    mbedtls_ssl_context ssl;
+    mbedtls_ssl_config conf;
+    mbedtls_x509_crt cacert;
+    uint32_t flags;
+    const char *pers = "ziti-sdk-c";
+
+    // Initialize the RNG and the session data
+    mbedtls_net_init( &server_fd );
+    mbedtls_ssl_init( &ssl );
+    mbedtls_ssl_config_init( &conf );
+    mbedtls_x509_crt_init( &cacert );
+    mbedtls_ctr_drbg_init( &ctr_drbg );
+
+    mbedtls_entropy_init( &entropy );
+    if( ( ret = mbedtls_ctr_drbg_seed( &ctr_drbg, mbedtls_entropy_func, &entropy, (const unsigned char *) pers, strlen( pers ) ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ctr_drbg_seed returned -0x%04x", -ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    struct http_parser_url controller_url;
+    http_parser_url_init(&controller_url);
+    http_parser_parse_url(ecfg->zej->controller, strlen(ecfg->zej->controller), 0, &controller_url);
+    char host[128];
+    char port[16];
+    int hostlen = controller_url.field_data[UF_HOST].len;
+    int hostoffset = controller_url.field_data[UF_HOST].off;
+    snprintf(host, sizeof(host), "%*.*s", hostlen, hostlen, ecfg->zej->controller + hostoffset);
+    sprintf(port, "%d", controller_url.port);
+
+    // Start the connection
+    ZITI_LOG(DEBUG, "Connecting to tcp/%s/%s...", host, port);
+    if( ( ret = mbedtls_net_connect( &server_fd, host, port, MBEDTLS_NET_PROTO_TCP ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_net_connect returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    if( ( ret = mbedtls_ssl_config_defaults( &conf, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ssl_config_defaults returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    mbedtls_ssl_conf_authmode( &conf, MBEDTLS_SSL_VERIFY_NONE );
+    mbedtls_ssl_conf_ca_chain( &conf, &cacert, NULL );
+    mbedtls_ssl_conf_rng( &conf, mbedtls_ctr_drbg_random, &ctr_drbg );
+    // mbedtls_ssl_conf_dbg( &conf, my_debug, stdout );
+
+    if( ( ret = mbedtls_ssl_setup( &ssl, &conf ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ssl_setup returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    if( ( ret = mbedtls_ssl_set_hostname( &ssl, host ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ssl_set_hostname returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    mbedtls_ssl_set_bio( &ssl, &server_fd, mbedtls_net_send, mbedtls_net_recv, NULL );
+
+    // Handshake
+    ZITI_LOG(DEBUG, "Performing the SSL/TLS handshake...");
+    while( ( ret = mbedtls_ssl_handshake( &ssl ) ) != 0 ) {
+        if( ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE ) {
+            ZITI_LOG(ERROR, "mbedtls_ssl_handshake returned -0x%x\n\n", -ret);
+            return ZITI_JWT_VERIFICATION_FAILED;
+        }
+    }
+
+    // Verify the server certificate
+    ZITI_LOG(DEBUG, "Verifying peer X.509 certificate...");
```

Comment:
Wouldn't the TLS handshake verify the server's cert?

Reply:
I have read that mbedtls_ssl_get_peer_cert() can sometimes fail after mbedtls_ssl_handshake() succeeds, and that calling mbedtls_ssl_get_verify_result() yields clues about what is wrong.